pywr-core/src/recorders/aggregator.rs

```diff
@@ -176,44 +204,127 @@ impl PeriodValue {
-            if av.is_some() {
-                if agg_value.is_some() {
-                    panic!("Multiple aggregated values yielded from aggregator. This indicates that the given value spans multiple aggregation periods which is not supported.")
+        if let Some(period) = &self.frequency {
+            for v in period.split_value_into_periods(value) {
+                let av = current_state.process_value(v, period, &self.function);
+                if av.is_some() {
+                    if agg_value.is_some() {
+                        panic!("Multiple aggregated values yielded from aggregator. This indicates that the given value spans multiple aggregation periods which is not supported.")
+                    }
+                    agg_value = av;
                 }
-                agg_value = av;
             }
+        } else {
+            current_state.process_value_no_period(value);
         }
-
         agg_value
     }
+
+    fn calc_aggregation(&self, state: &PeriodicAggregatorState) -> Option<PeriodValue> {
+        state.calc_aggregation(&self.function)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct AggregatorState {
+    state: PeriodicAggregatorState,
+    child: Option<Box<AggregatorState>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Aggregator {
+    agg: PeriodicAggregator,
+    child: Option<Box<Aggregator>>,
+}
+
+impl Aggregator {
+    pub fn new(period: Option<AggregationFrequency>, function: AggregationFunction, child: Option<Aggregator>) -> Self {
+        Self {
+            agg: PeriodicAggregator {
+                frequency: period,
+                function,
+            },
+            child: child.map(Box::new),
+        }
+    }
+
+    pub fn setup(&self) -> AggregatorState {
+        AggregatorState {
+            state: self.agg.setup(),
+            child: self.child.as_ref().map(|c| Box::new(c.setup())),
+        }
+    }
+
+    /// Append a new value to the aggregator.
+    pub fn append_value(&self, state: &mut AggregatorState, value: PeriodValue) -> Option<PeriodValue> {
+        let agg_value = match (&self.child, state.child.as_mut()) {
+            (Some(child), Some(child_state)) => child.append_value(child_state, value),
+            (None, None) => Some(value),
+            (None, Some(_)) => panic!("Aggregator state contains a child state when none is expected."),
+            (Some(_), None) => panic!("Aggregator state does not contain a child state when one is expected."),
+        };
+
+        if let Some(agg_value) = agg_value {
+            self.agg.process_value(&mut state.state, agg_value)
+        } else {
+            None
+        }
+    }
+
+    /// Compute the final aggregation value from the current state.
+    ///
+    /// This will also compute the final aggregation value from the child aggregators if any exists.
+    /// This includes aggregation calculations over partial or unfinished periods.
```

Comment:
Should this be a user setting?

Reply:
I'll add this as a new issue for the sake of getting this PR done.